cmake/modules/CheckCompiler.cmake

```diff
@@ -88,6 +88,7 @@ endif()
 
 if(NOT CMAKE_BUILD_TYPE)
   set(CMAKE_BUILD_TYPE RelWithDebInfo CACHE STRING "Choose the type of build, options are: Release, MinSizeRel, Debug, RelWithDebInfo." FORCE)
+  message(STATUS "No CMAKE_BUILD_TYPE set. Defaulting to ${CMAKE_BUILD_TYPE}.")
 endif()
```

Comment:
I think that this `if()` should be actually just removed. CMake should not add flags if the build type is not set. This default is bad for two reasons: 1) Setting a build type by default causes the flags for it to be appended to the user flags, which makes user-provided flags become ignored, and 2) in multi-configuration build systems, like with visual studio, the build type has no meaning at all and can mess up the configuration in other parts of the build system. @Axel-Naumann Do you have objections to removing this?

Reply:
I think we should be able to answer "what build do I get if I don't specify the build type" in a simple, robust way. What would that answer be without this `if`?

Reply:
What about checking if the user sets the CMAKE_CXX_FLAGS or the build type? If none are set, default the CXX_FLAGS to `-O2 -g`?
We would not interfere with custom configurations, but users that don't do anything get a sane default.

Reply:
You get a build with no extra flags added, i.e. the build type is unset. For multi-configuration build systems, no matter if it's set or not, the value of `CMAKE_BUILD_TYPE` is not used by CMake, and if *we* use it, we may get inconsistent results depending on which configuration we build. I had some trouble with the builtin PCRE due to that in the past. I removed a similar `if` from VecCore due to this as well. I think the important question is how to make ROOT respect the user's compilation flags, since now it's not doing that.

Reply:
Another important aspect is to not break the build of ROOT :-) So if we destroy the contract that `CMAKE_BUILD_TYPE` is always set we also need to remove all the places where it's expected. I know that e.g. cpack expects it.

Is this really in the scope of this PR? My suggestion: leave this as proposed by Stephan and go from there, after this PR has been merged (assuming the rest of this PR is okay).